stratum/lib/security/certificate.cc

```diff
@@ -148,9 +278,106 @@ util::Status Certificate::GenerateKeyPair(int bits) {
+      RETURN_ERROR(ERR_INVALID_PARAM) << "Invalid name param.";
+    case -1:  // internal error
+    default:
+      RETURN_ERROR(ERR_INTERNAL) << "OpenSSL error while checking common name.";
+  }
+}
+
+const std::string& Certificate::GetCommonName() { return common_name_; }
+
+util::Status Certificate::CheckIssuer(const Certificate& issuer) {
+  // Verify fields in the certificate are consistent (but not signatures)
+  int ret;
+  if ((ret = X509_check_issued(issuer.certificate_.get(),
+                               certificate_.get())) != X509_V_OK) {
+    RETURN_ERROR(ERR_INTERNAL) << "Issuer and cert do not match: "
+                               << X509_verify_cert_error_string(ret);
+  }
+
+  // Check that signature of this certificate was signed by the issuer
+  RETURN_IF_ERROR(
+      CheckSignatures(certificate_.get(), issuer.certificate_.get()));
+
+  RETURN_OK();
+}
+
+util::Status Certificate::LoadCertificate(const std::string& cert_pem,
+                                          const std::string& key_pem) {
+  // Load private key from PEM string
+  ASSIGN_OR_RETURN(bssl::UniquePtr<BIO> key_bio, StringToBio(key_pem));
+  bssl::UniquePtr<RSA> rsa;
+  OPENSSL_RETURN_IF_ERROR(rsa = bssl::UniquePtr<RSA>(PEM_read_bio_RSAPrivateKey(
+                              key_bio.get(), nullptr, nullptr, nullptr)));
+  // Store this keypair in EVP
+  OPENSSL_RETURN_IF_ERROR(EVP_PKEY_assign_RSA(key_.get(), rsa.get()));
+  // RSA will be freed when EVP is freed, so release the pointer to avoid
+  // deletion. Note: RSA will still be deleted by the shared pointer on error.
+  rsa.release();
+
+  // Load certificate from PEM string
+  ASSIGN_OR_RETURN(bssl::UniquePtr<BIO> cert_bio, StringToBio(cert_pem));
+  OPENSSL_RETURN_IF_ERROR(certificate_ =
+                              bssl::UniquePtr<X509>(PEM_read_bio_X509(
+                                  cert_bio.get(), nullptr, nullptr, nullptr)));
+
+  // Copy common name from loaded certificate
+  X509_NAME* name;
+  OPENSSL_RETURN_IF_ERROR(name = X509_get_subject_name(certificate_.get()));
+  bssl::UniquePtr<BIO> name_bio = NewBio();
+  OPENSSL_RETURN_IF_ERROR(
+      X509_NAME_print_ex(name_bio.get(), name, 0,
+                         ASN1_STRFLGS_RFC2253 | ASN1_STRFLGS_ESC_QUOTE |
+                             XN_FLAG_SEP_CPLUS_SPC | XN_FLAG_FN_NONE));
+  common_name_ = BioToString(name_bio);
+
+  // Check that the certificate and private key match
+  ERR_clear_error();
+  switch (X509_check_private_key(certificate_.get(), key_.get())) {
+    case 1:  // cert and key match
+      break;
+    case 0:  // cert and key do not match
+      RETURN_ERROR(ERR_INVALID_PARAM)
+          << "Certificate and private key do not match.";
+    default:
+      auto bio = NewBio();
+      ERR_print_errors(bio.get());
+      RETURN_ERROR(ERR_INVALID_PARAM) << "OpenSSL error checking private key.\n"
+                                      << BioToString(bio);
+  }
+
+  RETURN_OK();
```

Comment:
Did you pick this macro up from PINS? I have not seen any other use of `RETURN_OK` in Stratum.

Reply:
We don't have any uses of it, but it's been in the Stratum tree since the beginning:
https://github.com/stratum/stratum/blob/main/stratum/glue/status/status_macros.h#L383

I just happened to notice it and thought I'd try it out. Happy to switch back to:
```c++
return ::util::OkStatus();
```

In PINS, it's mostly: `return absl::OkStatus();`

Reply:
Interesting. I guess Google had to keep both around because of widespread use.

For us `::util::OkStatus()` would be more consistent, don't you agree?

Reply:
Yup, I agree. changed them over to `util::OkStatus()`